### ISSUE TITLE
Estimating data transfer costs

### ISSUE THREAD

**Issue by alikhajeh1** [infracost/infracost#220]
As we're getting close to supporting GCP prices (#24), it would be good to start discussing how we can estimate data transfer costs for AWS and GCP (we'll probably use the same mechanism when we add Azure).

Some options I can think of are:
1. Add one new resource to [the infracost terraform provider](https://github.com/infracost/terraform-provider-infracost) so users could input total amounts of data transfer for each type (from region to internet, from region to another region, cross-AZ...) and leave infracost to calculate cost estimates. This seems easy to implement but requires users to provide total usage numbers.
2. Specify data transfer properties on a per-resource basis, possibly using the existing resources (though most resources don't have infracost provider-support yet), and expect users to provide estimates for resources they want, and let infracost sum them up, then calculate cost estimates. This would take more effort to implement, and I'm not sure if users would know these numbers on a per-resource basis?

@aliscott, @shomeprasanjit, @robh007 any thoughts or other options?

**Comment by aliscott** [comment]
The current Infracost provider resources all have a `resources` attribute that specifies a list of resources that it applies to. We could use a similar concept for this - have a single new resource type for data transfer but use the `resources` attribute to associate it to a group of cloud resources. This way the user can either specify one or many of these data transfer resources, depending on how they want to associate them to existing resources.

**Comment by shomeprasanjit** [comment]
As per our use-case, specifying data transfer properties per-resource does give us the flexibility to quantify and single-out cost per resource basis especially needed during drill down of infrastructure cost.  I think we should think about some mechanism to pass a template to the infracost utility containing specific pre-defined keys (ex: requests/sec, data transfer volume(in/out) etc based on per day/per hour/per month)  and we just fill in the values to it. lastly have infracost to calculate cost estimates. 

One more point, I think we should also come up with some default values incase these values are not passed by users and call out those values while showing the output. 

let me know your thoughts.

**Comment by alikhajeh1** [comment]
@shomeprasanjit good points, we're seeing the need to have something like an infracost config file so the things you discuss could possibly be set there. As a first cut, if I understand @aliscott's proposal, users could add something like the following to specify per-resource as well as project-wide data transfer properties. These would be summed by infracost and multiplied by the tiered pricing to calculate the total cost.

```
data "infracost_aws_data_transfer" "my_resource_data_transfer" {
  resources = list(aws_apigatewayv2_api.my_http.id, aws_apigatewayv2_api.my_websocket.id)

  monthly_data_out {
    value = 10 # GB
  }
}

data "infracost_aws_data_transfer" "other_data_transfer" {
  monthly_data_out {
    value = 5 # GB
  }
}
```

Regarding default values, I think it's worth exploring, maybe as a config option that could be turned on/off, since there are many usage-based resources and infracost showing `-` for their cost might look odd for first-time users.

**Comment by alikhajeh1** [comment]
Should data transfer prices always be shown in the infracost output since it's a "global" price? the AWS pricing pages always seem to include data transfer prices for all resources.

**Comment by shomeprasanjit** [comment]
AFAIK, data transfer cost is usually associated with individual resources. hence should be included in the drill down list of total aggregated cost under each resource.

**Comment by alikhajeh1** [comment]
Thanks @shomeprasanjit! Let's assume the following setup:
```
data "infracost_aws_data_transfer" "my_resource_data_transfer" {
  resources = list(aws_apigatewayv2_api.my_http.id, aws_apigatewayv2_api.my_websocket.id) # Would be great if we could apply wildcards here too, e.g. 'aws_apigatewayv2_api.my_app1*'

  monthly_data_out {
    value = 10 # GB
  }
}

data "infracost_aws_data_transfer" "other_data_transfer" {
  monthly_data_out {
    value = 5 # GB
  }
}
```

Since data transfer is tiered, and can be used by most cloud resources, I can think of 3 options for showing it in the infracost output:
1. Show one `data transfer` cost component **per** cloud resource that uses `infracost_aws_data_transfer`. Use the first non-free tier to calculate the cost. The downside is that tiered prices are ignored, thus we might be over-estimating the costs.
2. Show **one** `data transfer` cost component at the bottom of the infracost output, sum all of the `infracost_aws_data_transfer` values and calculate the cost using the applicable tiers. The downside is that users won't be able to see the data transfer cost of each resource.
3. Hybrid of 1 and 2: sum all of the `infracost_aws_data_transfer` values and calculate the total cost using tiered pricing, then calculate the user's per-GB data transfer price. Show a cost component **per** cloud resource that uses `infracost_aws_data_transfer`, as well as **one** at the end for any non-resource-specific data transfer.

Option 3 is the most complicated to implement but it seems the best for users?

**Comment by shomeprasanjit** [comment]
@alikhajeh1, Infracost provides us "before the fact" pricing and hence meant to be ballpark estimates only, so I won't be too concerned about over-estimating.  AFAIC, Option 3 is the best way to describe both resource-specific and non-resource specific (outbound, in-bound traffic) data transfer costs. 
And just to be clear we are dealing with on-demand prices only and doesn't include any Savings Plan, right?

**Comment by alikhajeh1** [comment]
@shomeprasanjit yep, initially just on-demand pricing. Longer term we could connect to the users cloud account and take into account Savings Plans, RIs and any custom price books.

**Comment by sinabakh** [comment]
I think I might be a bit out of context. As far as I know, we always ignore the free tier in our pricing. So why should we handle it for data transfer costs?

**Comment by alikhajeh1** [comment]
@sinabakh we're ignoring the free tier for now for data transfer, I updated https://www.infracost.io/docs/supported_resources just now to mention that "Most expensive price tier is used".

@shomeprasanjit https://github.com/infracost/infracost/blob/master/infracost-usage-example.yml#L24-L29 shows an example of how the data transfer can be defined, each block `aws_data_transfer.my_resource_1` or `aws_data_transfer.my_region` will show-up separately in the infracost output. We decided not to implement the tiered price calculation so we can ship the usage-file feature to gather feedback on the approach quicker. We'll get round to calculating price tiers more accurately once users are happy with the usage-based resources :)

Going to close this issue, feedback via comments here or the community Slack channel is welcome. New issues can be created for bugs or other ideas on improving this.
